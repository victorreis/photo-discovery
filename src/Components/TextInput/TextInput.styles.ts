import styled from 'styled-components';

import { toPx } from '../../Utils/Transform/toPx.util';

export const Input = styled.input`
  height: 3rem;
  margin: 0.5rem;
  border-width: 1px;
  border-color: ${({ theme }) => theme.colors.main.effect.normal}55;
  border-style: solid;
  ${({ theme }) => theme.typographies.body1};
  background-color: ${({ theme }) => theme.colors.background.default.lightest};
  border-radius: ${({ theme }) => toPx(theme.borders.radius.LG)};
  box-sizing: border-box;
  padding: 0.5rem;
`;
