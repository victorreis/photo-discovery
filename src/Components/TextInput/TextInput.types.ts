import { TestProps } from '../../Config/Tests/Test.types';

export interface RequiredTextInputProps {
  /**
   * Value to be controlled by the component.
   */
  value: string;

  /**
   * Callback function that is called when the input is changed.
   */
  onChange: (newValue: string) => void;
}

export interface DefaultTextInputProps {}

export interface OptionalTextInputProps {
  /**
   * Regular expression to be used to filter the input data.
   */
  filterInputRegex?: RegExp;
}

export type TextInputProps = RequiredTextInputProps &
  DefaultTextInputProps &
  OptionalTextInputProps &
  TestProps &
  Omit<React.HTMLAttributes<HTMLInputElement>, 'onChange'>;

export type TextInputStyleProps = Required<DefaultTextInputProps>;
