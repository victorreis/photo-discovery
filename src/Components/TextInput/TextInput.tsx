import { useState, forwardRef, ForwardRefRenderFunction } from 'react';

import { TestProps } from '../../Config/Tests/Test.types';
import { Input } from './TextInput.styles';
import { TextInputProps, DefaultTextInputProps } from './TextInput.types';

export const textInputDefaults: Required<DefaultTextInputProps> &
  Required<TestProps> = {
  testID: 'TextInput',
};

const TextInputComponent: ForwardRefRenderFunction<
  HTMLInputElement,
  TextInputProps
> = (props, ref) => {
  const {
    testID = textInputDefaults.testID,
    children,
    value: initialValue,
    onChange,
    filterInputRegex,
    ...others
  } = props;

  const [value, setValue] = useState(initialValue);

  const handleChange = (event: React.ChangeEvent<HTMLInputElement>) => {
    let newValue = event.target.value;
    if (filterInputRegex) {
      newValue = newValue.replaceAll(filterInputRegex, '');
    }
    setValue(() => newValue);
    onChange(newValue);
  };

  return (
    <Input
      data-testid={testID}
      onChange={handleChange}
      ref={ref}
      type="text"
      value={value}
      {...others}
    >
      {children}
    </Input>
  );
};
export const TextInput = forwardRef(TextInputComponent);
