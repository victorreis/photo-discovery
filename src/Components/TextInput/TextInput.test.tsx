import {
  renderJestDomCreator,
  renderRTRCreator,
  screen,
} from '../../Config/Tests/GlobalSetup.config';
import { TextInput, textInputDefaults } from './TextInput';
import { RequiredTextInputProps, TextInputProps } from './TextInput.types';

describe(`TextInput component tests`, () => {
  const value = 'text';
  const onChange = jest.fn();

  const requiredProps: RequiredTextInputProps = {
    value,
    onChange,
  };

  const setup = (props?: TextInputProps) => {
    const renderRTR = () =>
      renderRTRCreator<TextInputProps>(TextInput, {
        ...requiredProps,
        ...props,
      });
    const renderJestDom = () =>
      renderJestDomCreator<TextInputProps>(TextInput, {
        ...requiredProps,
        ...props,
      });

    return { renderRTR, renderJestDom };
  };

  describe(`behavior tests`, () => {
    it(`should render the component`, () => {
      setup().renderJestDom();
      const testInstance = screen.getByTestId(textInputDefaults.testID);

      expect(testInstance).toBeTruthy();
    });

    // it(`should render the text`, () => {
    //   setup().renderJestDom();
    //   const element = screen.getByText(text);

    //   expect(element).toBeInTheDocument();
    // });

    // it(`should render '${textInputDefaults.variant}' as the default variant`, () => {
    //   const instance = setup().renderRTR().root;
    //   const element = instance.findByProps({
    //     variant: textInputDefaults.variant,
    //   });

    //   expect(element).toBeTruthy();
    // });

    // it(`should override the default variant when it is passed as prop`, () => {
    //   const instance = setup({
    //     ...requiredProps,
    //     variant: newVariant,
    //   }).renderRTR().root;
    //   const element = instance.findByProps({ variant: newVariant });

    //   expect(element).toBeTruthy();
    // });
  });

  // describe(`style tests`, () => {
  //   it(`should have style the Container component`, () => {
  //     setup().renderJestDom();
  //     const container = screen.getByTestId(textInputDefaults.testID);

  //     expect(container).toHaveStyle({
  //       color: hexToRgb(themes.default.colors.font.default),
  //       ...themes.default.typographies[textInputDefaults.variant],
  //     });
  //   });
  // });

  describe(`snapshot tests`, () => {
    it(`should render correctly`, () => {
      const generatedJson = setup().renderRTR().toJSON();
      expect(generatedJson).toMatchSnapshot();
    });
  });
});
