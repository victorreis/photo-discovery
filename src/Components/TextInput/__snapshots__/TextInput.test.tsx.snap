// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextInput component tests snapshot tests should render correctly 1`] = `
<input
  className="Input-photo-discovery__sc-oh0frq-0 egkOIw"
  data-testid="TextInput"
  onChange={[Function]}
  type="text"
  value="text"
/>
`;
